tests/defences/detector/poison/test_clustering_centroid_analysis.py

```diff
@@ -0,0 +1,1442 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2018
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import json
+import logging
+import unittest
+from unittest.mock import MagicMock, patch
+
+import numpy as np
+import pytest
+import tensorflow as tf
+from tensorflow import keras
+from sklearn.base import ClusterMixin
+from tensorflow.keras import Model, Sequential, Input
+from tensorflow.keras.layers import Dense
+from tensorflow.keras.losses import CategoricalCrossentropy
+from tensorflow.keras.optimizers import Adam
+
+from art.defences.detector.poison.clustering_centroid_analysis import (
+    ClusteringCentroidAnalysis,
+    _calculate_centroid,
+    _class_clustering,
+    _feature_extraction,
+    _cluster_classes,
+    _encode_labels,
+)
+from art.estimators.classification import TensorFlowV2Classifier
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+class MockClusterer(ClusterMixin):
+    """
+    A mock ClusterMixin for testing purposes.  This avoids using a real clustering
+    algorithm and allows us to control the output for our tests.
+    """
+
+    def __init__(self, cluster_labels_to_return):
+        self.cluster_labels_to_return = cluster_labels_to_return
+
+    def fit_predict(self, x, **kwargs):
+        return self.cluster_labels_to_return
+
+
+@pytest.mark.skip_framework(
+    "tensorflow1",
+    "keras",
+    "kerastf",
+    "pytorch",
+    "mxnet",
+    "non_dl_frameworks",
+)
```

Comment:
We are removing TF1 and MXNet in ART 1.20.
```suggestion
@pytest.mark.skip_framework(
    "keras",
    "kerastf",
    "pytorch",
    "non_dl_frameworks",
)
```